### PR TITLE
Cluster creation refactor

### PR DESCRIPTION
**Pull request by ubergesundheit**
This PR refactors a huge portion of cluster creation and node pool creation. For now this focues (mostly) on untangling the css shared between some components.

I hope this PR isn't too big. If you want we could go through the changes in a call or I could split this up.

Towards giantswarm/giantswarm#11004 giantswarm/giantswarm#11082

Some notable changes:

- Introduction of several UI components derived from cluster & node pool creation
- Introduction of a "base" spacing which can be used to make margins & paddings more consistent ([Definition](https://github.com/giantswarm/happa/pull/1544/files#diff-ca70e333d313532b406dbb1fb36766c6R65), [Usage example](https://github.com/giantswarm/happa/pull/1544/files#diff-b1d40bbe9324d38f81b881636f301e86R4-R10))
- Usage of the our `<Input>` component for cluster name and node pool name as well as wrapping other inputs like release selector and instance type selector

Here are some screenshots:

## AWS v4 Cluster creation

<details>
<summary>Click here for screenshots</summary>

### Current
aws_legacy_old

### This PR
aws_legacy_new

</details>

## AWS v5 Cluster creation (node pools)

<details>
<summary>Click here for screenshots</summary>

### Current
aws_v5_old

### This PR
aws_v5_new

</details>

## AWS v5 Cluster creation (node pools) \# 2

<details>
<summary>Click here for screenshots</summary>

### Current
aws_v5_2_old

### This PR
aws_v5_2_new

</details>

## AWS add node pool from cluster detail page

<details>
<summary>Click here for screenshots</summary>

### Current
aws_add_np_new

### This PR
aws_add_np_old

</details>

## Azure < 11.1.0

<details>
<summary>Click here for screenshots</summary>

### Current
azure_pre_old

### This PR
azure_pre_new

</details>

## Azure > 11.1.0

<details>
<summary>Click here for screenshots</summary>

### Current
azure_post_old

### This PR
azure_post_new

</details>

## KVM

<details>
<summary>Click here for screenshots</summary>

### Current
kvm_old

### This PR
kvm_new

</details>

### CODE REVIEW

**Comment by axbarsan** [comment]
> This PR refactors a huge portion of cluster creation and node pool creation. For now this focues (mostly) on untangling the css shared between some components.
> 
> I hope this PR isn't too big. If you want we could go through the changes in a call or I could split this up.
> 
> Towards [giantswarm/giantswarm#11004](https://github.com/giantswarm/giantswarm/issues/11004) [giantswarm/giantswarm#11082](https://github.com/giantswarm/giantswarm/issues/11082)
> 
> Some notable changes:
> 
> * Introduction of several UI components derived from cluster & node pool creation
> * Introduction of a "base" spacing which can be used to make margins & paddings more consistent ([Definition](https://github.com/giantswarm/happa/pull/1544/files#diff-ca70e333d313532b406dbb1fb36766c6R65), [Usage example](https://github.com/giantswarm/happa/pull/1544/files#diff-b1d40bbe9324d38f81b881636f301e86R4-R10))
> * Usage of the our `<Input>` component for cluster name and node pool name as well as wrapping other inputs like release selector and instance type selector
> 
> Here are some screenshots:
> 
> ## AWS v4 Cluster creation
> Click here for screenshots
> ## AWS v5 Cluster creation (node pools)
> Click here for screenshots
> ## AWS v5 Cluster creation (node pools) # 2
> Click here for screenshots
> ## AWS add node pool from cluster detail page
> Click here for screenshots
> ## Azure < 11.1.0
> Click here for screenshots
> ### Current
> azure_pre_old
> 
> ### This PR
> azure_pre_new
> 
> ## Azure > 11.1.0
> Click here for screenshots
> ### Current
> azure_post_old
> 
> ### This PR
> azure_post_new
> 
> ## KVM
> Click here for screenshots
> ### Current
> kvm_old
> 
> ### This PR
> kvm_new

**Comment by axbarsan** [comment]
Sorry for that, fat fingers 😓

File: src/components/Cluster/NewCluster/CreateNodePoolsCluster.js [hunk 1]
```diff
@@ -359,212 +220,160 @@ class CreateNodePoolsCluster extends Component {
...
+              {hasAZLabels && zonesArray.length > maxAZ && (
+                <p className='danger'>
+                  {maxAZ} is the maximum you can have. Please uncheck{' '}
+                  {zonesArray.length - maxAZ} of them.
+                </p>
+              )}
+            </AZWrapperDiv>
+          </FlexColumn>
+          {Object.keys(nodePools).length === 0 && <HorizontalLine />}
+          <NodePoolsTransitionGroup>
```
Comment (axbarsan) [inline src/components/Cluster/NewCluster/CreateNodePoolsCluster.js:292]: Do you think we can extract the Node Pools functionality to another component?

File: src/components/Cluster/NewCluster/CreateRegularCluster.js [hunk 2]
```diff
@@ -408,213 +380,177 @@ class CreateRegularCluster extends React.Component {
...
+              switch (provider) {
+                case Providers.AWS: {
+                  const [RAM, CPUCores] = this.produceRAMAndCoresAWS();
+
+                  return (
+                    <StyledInput
+                      inputId='instance-type'
+                      label='Instance type'
+                      // regular space, hides hint ;)
+                      hint={<>&#32;</>}
```
Comment (axbarsan) [inline src/components/Cluster/NewCluster/CreateRegularCluster.js:408]: Sneaky 😬

File: src/components/Cluster/NewCluster/CreateRegularCluster.js [hunk 3]
```diff
@@ -408,213 +380,177 @@ class CreateRegularCluster extends React.Component {
...
+          {(provider === Providers.AWS || provider === Providers.AZURE) && (
+            <V4AvailabilityZonesSelector
+              minValue={this.props.minAvailabilityZones}
+              maxValue={this.props.maxAvailabilityZones}
+              onChange={this.updateAvailabilityZonesPicker}
+              {...multiAZSelectorProps}
+            />
+          )}
+          <label htmlFor='instance-type' className='instance-type'>
+            {(() => {
```
Comment (axbarsan) [inline src/components/Cluster/NewCluster/CreateRegularCluster.js:398]: Would be awesome if we could move this switch statement into a function, and extract each switch case to a separate component. What do you think?

File: src/components/UI/NumberPicker.js [hunk 4]
```diff
@@ -222,7 +223,7 @@ class NumberPicker extends React.Component {
   render() {
     return (
       <Wrapper
-        className={`${this.props.theme ? this.props.theme : ''} ${
+        className={`${this.props.className ?? ''} ${this.props.theme ?? ''} ${
```
Comment (axbarsan) [inline src/components/UI/NumberPicker.js:226]: Not absolutely necessary for this PR, but maybe we should create a helper function for classnames at some point.

I'm thinking of something like
```typescript
function classNames(...classes: Array<string | undefined>): string {
...join into one string, ignoring undefined values
}
```

File: src/components/UI/ClusterCreation/InstanceTypeSelection.tsx [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import styled from '@emotion/styled';
+
+import { AdditionalInputHint } from './StyledInput';
+
+export const InstanceTypeDescription = styled(AdditionalInputHint)`
+  margin: 0 0 0
+    ${({ theme }) =>
+      /* eslint-disable-next-line no-magic-numbers */
+      theme.spacingPx * 4}px;
```
Comment (axbarsan) [inline src/components/UI/ClusterCreation/InstanceTypeSelection.tsx:9]: Not a big fan of this notation, I'd prefer to have all margin sizes on the same line, but not a blocker

File: src/components/Cluster/NewCluster/CreateNodePoolsCluster.js (hunk 1 above)
Comment (ubergesundheit) [reply src/components/Cluster/NewCluster/CreateNodePoolsCluster.js:292]: Sure, but not in this PR ;)

File: src/components/Cluster/NewCluster/CreateRegularCluster.js (hunk 3 above)
Comment (ubergesundheit) [reply src/components/Cluster/NewCluster/CreateRegularCluster.js:398]: Agree, I can do it in the same PR where I do this: https://github.com/giantswarm/happa/pull/1544#discussion_r437637115

File: src/components/UI/NumberPicker.js (hunk 4 above)
Comment (ubergesundheit) [reply src/components/UI/NumberPicker.js:226]: Maybe use this: https://github.com/lukeed/clsx

Comment (axbarsan) [reply src/components/UI/NumberPicker.js:226]: C'mon man, we can do that in 3 lines of code, no need for a library 😬

File: src/components/UI/ClusterCreation/InstanceTypeSelection.tsx (hunk 5 above)
Comment (ubergesundheit) [reply src/components/UI/ClusterCreation/InstanceTypeSelection.tsx:9]: yeah..

> Looks like the magic number linter rule is making our code really dirty. Is it maybe time to disable it?

Fine for me. We could ignore 1-9 in the rule config for starters.

**Comment by axbarsan** [comment]
This is how cluster creation worker nodes section looks for me on Azure (godsmack)

image

The problem is that we're rendering the text field label, even though it's empty, and that has a bottom margin.

image

**Comment by axbarsan** [comment]
^ Scratch that, I should learn how Kubernetes works 🤦

**Pull request merged**